### PR TITLE
[GH-855] Save user's last used field values

### LINKED ISSUE #855: Save user's last used field values from Create Issue modal, to be used for future issue creations
At the moment, the plugin saves a given user's last selected Jira project in the Create Issue modal. However other fields, such as issue type and labels, are not saved.

The task here is to make it so when a user submits the Create Issue modal, the field values in the submission can be used for creating another ticket in the future. A more advanced version of this feature allow for custom configurations of field values, but the scope of this task is just to support the last used form submission.

Relevant thread https://community-daily.mattermost.com/core/pl/pxdo44fnk7rkxf1tmqge65g9rw

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
#### Summary
Updated logic to store the last used issue type value while creating the issue and prefill the modal with those values the next time the user creates an issue.

##### What should be tested?
1. The specified fields are getting pre-filled properly.
2. A new issue is getting created successfully.
3. The prefilled value does not exist on Jira anymore.

#### Ticket Link
Fixes #855

### CODE REVIEW

File: server/user.go [hunk 1]
```diff
@@ -33,8 +33,14 @@ type Connection struct {
 	Oauth1AccessSecret string        `json:",omitempty"`
 	OAuth2Token        *oauth2.Token `json:",omitempty"`
 	Settings           *ConnectionSettings
-	DefaultProjectKey  string   `json:"default_project_key,omitempty"`
-	MattermostUserID   types.ID `json:"mattermost_user_id"`
+	SavedFieldValues   *SavedFieldValues `json:"saved_field_values,omitempty"`
```
Comment (mickmister) [inline server/user.go:36]: We are losing data here by changing its shape, but the "old" value of `DefaultProjectKey` is not too important. Just something to think about when changing persistent data structures

File: webapp/src/components/modals/create_issue/create_issue_form.tsx [hunk 2]
```diff
@@ -132,21 +133,32 @@ export default class CreateIssueForm extends React.PureComponent<Props, State> {
...
+            };
+        } else {
+            const issueTypes = getIssueTypes(this.state.jiraIssueMetadata, projectKey);
+            const issueType = issueTypes.length ? issueTypes[0].id : '';
+            fields.issuetype = {
+                id: issueType,
+            };
+        }
+
+        fields = {...fields};
```
Comment (mickmister) [inline webapp/src/components/modals/create_issue/create_issue_form.tsx:154]: Does this need to be here?

**Review by mickmister — APPROVED** [review]

**Comment by mickmister** [comment]
So the one field we save is the "component" type. I'm thinking the point of the feature is to fill in more fields than just the components. From the discussion at https://github.com/mattermost/mattermost-plugin-jira/issues/855, we didn't land on a definitive requirement to fulfill here. @matthewbirtch Any thoughts on how exactly this feature should work?

**Comment by mickmister** [comment]
One thing I suggested in that discussion is having a "Clear All" button that allows you to start from scratch if the filled out fields don't apply to the second ticket you're creating

https://github.com/mattermost/mattermost-plugin-jira/issues/855#issuecomment-1121192599

**Comment by matthewbirtch** [comment]
> So the one field we save is the "component" type. I'm thinking the point of the feature is to fill in more fields than just the components. From the discussion at #855, we didn't land on a definitive requirement to fulfill here. @matthewbirtch Any thoughts on how exactly this feature should work?
> 
> Personally I think we should save everything that's not the summary nor a textarea (description or custom textarea), which seems to be what Jira does with their "Create another issue" checkbox feature

I'm not sure we should prefill that much. I think we should prefill only the following based on the previous selection:
- Project
- Issue Type
- Components
- Team

I don't imagine Fix Version, Epic, Assignee, Labels, etc would be repeated as much from one ticket to the next

**Comment by mickmister** [comment]
@matthewbirtch "Team" is not a built-in Jira field. Instead, it's a custom field we have on our Jira instance. Unfortunately we don't have a way to give precedence of that field over other custom fields, unless we make it so this can be explicitly configured somewhere. The plugin already supports the last selected "project", and this feature is currently adding functionality for "issue type" and "components".

I personally don't use components that often (though I know they are commonly used), and I think auto-selecting "issue type" wouldn't save me much time when creating an issue. This feature request came up for use during a bug bash, where many of the fields would be the same for each ticket. That's where I see it being most useful, and saving the most clicks in the form.

**Comment by matthewbirtch** [comment]
> @matthewbirtch "Team" is not a built-in Jira field. Instead, it's a custom field we have on our Jira instance. Unfortunately we don't have a way to give precedence of that field over other custom fields, unless we make it so this can be explicitly configured somewhere. The plugin already supports the last selected "project", and this feature is currently adding functionality for "issue type" and "components".
> 
> I personally don't use components that often (though I know they are commonly used), and I think auto-selecting "issue type" wouldn't save me much time when creating an issue. This feature request came up for use during a bug bash, where many of the fields would be the same for each ticket. That's where I see it being most useful, and saving the most clicks in the form.

I'm not sure the bug bash use case is ubiquitous enough to save all fields from previous issue created with the plugin. Maybe we could assume you want that if you're creating issues in quick succession, but if I create an issue today and then create the next issue tomorrow, I would find it confusing why all the fields were prefilled and I would likely have to clear most of them which actually makes it take longer.

**Comment by mickmister** [comment]
> I would likely have to clear most of them which actually makes it take longer.

@matthewbirtch This is where the "Clear all" button would be useful, though it's really just a bandaid on the issue

Thinking about this some more, I think the pre-populating of form fields should be a bit more explicit. Any "implicit" approach (pre-filling Component without explicitly telling the user this is happening) can cause issues where the value was unknowingly and accidentally assigned a value, when it should have been blank.

We could have a button for the user to click when they open the modal that says "prefill fields with values from my last submission". This way the user has full control over the automation happening, and can decide on the spot what they want to happen in the form.

**Comment by matthewbirtch** [comment]
> > I would likely have to clear most of them which actually makes it take longer.
> 
> @matthewbirtch This is where the "Clear all" button would be useful, though it's really just a bandaid on the issue
> 
> Thinking about this some more, I think the pre-populating of form fields should be a bit more explicit. Any "implicit" approach (pre-filling Component without explicitly telling the user this is happening) can cause issues where the value was unknowingly and accidentally assigned a value, when it should have been blank.
> 
> We could have a button for the user to click when they open the modal that says "prefill fields with values from my last submission". This way the user has full control over the automation happening, and can decide on the spot what they want to happen in the form.

@mickmister I'm hesitant to add either of the 'clear all' or the 'prefill fields' buttons since they are not common, expected elements in a form. Reset buttons are [not recommended as a UX](https://www.nngroup.com/articles/reset-and-cancel-buttons/) best practice. I'd be a bigger proponent of prefilling a few key fields as sensible defaults and just leaving the rest out.

File: webapp/src/components/modals/create_issue/create_issue_form.tsx (hunk 2 above)
Comment (raghavaggarwal2308) [reply webapp/src/components/modals/create_issue/create_issue_form.tsx]: @mickmister Fixed

**Comment by mickmister** [comment]
> I'd be a bigger proponent of prefilling a few key fields as sensible defaults and just leaving the rest out.

@matthewbirtch I'm thinking the functionality of "autofill Component" could provide more trouble than good. Making the "issue type" autofilled does make sense as it's an obvious step in the form that's being filled out, but the `Component` field is more implicit and potentially hard to notice that was autofilled, resulting in potentially incorrect data being submitted.

I'm thinking we should decrease the scope of this task to only include `issue type`, and leave the `Component` out of the changes. What do you think?

**Comment by matthewbirtch** [comment]
> I'm thinking we should decrease the scope of this task to only include issue type, and leave the Component out of the changes. What do you think?

I'm good with that for now @mickmister

**Comment by raghavaggarwal2308** [comment]
@mickmister @matthewbirtch Removed the logic to store the component field.

**Comment by AayushChaudhary0001** [comment]
Tested and working fine for the following conditions:

- Issue type field prefilled correctly.
- New issue is being created.
- Issue type field is empty if the desired value (issue type) does not exist on Jira.

Tested for both Jira Cloud and Jira Server. Approved, LGTM.

**Review by AayushChaudhary0001 — APPROVED** [review]

**Comment by raghavaggarwal2308** [comment]
@mickmister Gentle reminder for re-review

**Comment by raghavaggarwal2308** [comment]
@mickmister Gentle reminder for re-review

**Review by mickmister — APPROVED** [review]

**Comment by hanzei** [comment]
@raghavaggarwal2308 Heads up that the PR broke `master`: https://github.com/mattermost/mattermost-plugin-jira/actions/runs/8276109706/job/22644115255

**Pull request merged**